### PR TITLE
Fixes for use of emcee v3 (remove PTsampler), and allow iter_cb to be run and abort fit

### PR DESCRIPTION
**Pull request by newville**
This replaces #594 and addresses #590 by explicitly requiring and supporting only version 3 of `emcee`.  
  
In working on this, I have come to the conclusion that fitting with `emcee` actually does not work very well.  So, although this PR allows code using `emcee 3` to run to completion, the results are rarely good.   I am not certain why this is and will raise as a separate issue.   

My sense is that we should merge this so that there is "working code" and then decide whether to try to fix or drop the use of `emcee` as a fitting method.

###### Type of Changes

- [x] Bug fix
- [ ] New feature
- [x] Refactoring / maintenance
- [ ] Documentation / examples

###### Tested on

Python: 3.7.3 (default, Mar 27 2019, 16:54:48) 
[Clang 4.0.1 (tags/RELEASE_401/final)]

lmfit: 0.9.14+43.gb8fb890, scipy: 1.3.0, numpy: 1.16.4, asteval: 0.9.15, uncertainties: 3.0.2, six: 1.12.0

###### Verification 
Have you

- [ ] included docstrings that follow PEP 257?

- [x] referenced existing Issue and/or provided relevant link to mailing list?

- [x] verified that existing tests pass locally?

- [ ] verified that the documentation builds locally?

- [ ] squashed/minimized your commits and written descriptive commit messages?

- [x] added or updated existing tests to cover the changes?
- [ ] updated the documentation?
- [ ] added an example?

### CODE REVIEW

File: lmfit/minimizer.py [hunk 1]
```diff
@@ -45,8 +45,7 @@
 try:
     import emcee
     from emcee.autocorr import AutocorrError
-    HAS_EMCEE = True
-    EMCEE_VERSION = int(emcee.__version__[0])
+    HAS_EMCEE = int(emcee.__version__[0]) >= 3
```
[the remark this answers is not in the document]
Comment (newville) [reply lmfit/minimizer.py]: well, a user may have installed emcee separately -- we don't require it, but if it is installed it has to be version 3

File: lmfit/minimizer.py [hunk 2]
```diff
@@ -66,6 +65,14 @@
 except ImportError:
     HAS_NUMDIFFTOOLS = False
 
+try:
+    import dill
+    dill_version = dill.__version__
```
[the remark this answers is not in the document]
Comment (newville) [reply lmfit/minimizer.py]: OK, I'll change that.  Either way of adding a little cruft to silence flake8 is ok with me.

File: requirements-dev.txt [hunk 3]
```diff
@@ -3,7 +3,7 @@ codecov
 corner
 coverage
 dill
-emcee<3.0.0
+emcee>=3.0.0
```
[the remark this answers is not in the document]
Comment (newville) [reply requirements-dev.txt]: emcee isn't in requirements.txt.  I think that's OK.   If anything, I prefer removing all usage of `emcee` -- it appears to be wrong to me, and quite possibly dangerously so.

File: tests/test_parameter.py [hunk 4]
```diff
@@ -526,13 +526,3 @@ def test__rsub__(parameter):
     """Test the __rsub__ magic method."""
     par, _ = parameter
     assert_allclose(5.25 - par, -4.75)
-
-
-def test_isParameter(parameter):
```
[the remark this answers is not in the document]
Comment (newville) [reply tests/test_parameter.py]: OK, I'll restore it.  I don't have a strong preference.    In fact, I'm ok with un-deprecating... it's not really doing any harm and some downstream project might want to know if an object is an `lmfit.Parameter`.

File: lmfit/minimizer.py [hunk 5]
```diff
@@ -1140,17 +1225,15 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
             raise NotImplementedError('You must have emcee to use'
                                       ' the emcee method')
```
[the remark this answers is not in the document]
Comment (newville) [reply lmfit/minimizer.py]: OK, will do.

File: lmfit/minimizer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (reneeotten) [reply lmfit/minimizer.py]: sorry, you're right

**Comment by newville** [comment]
I don't really understand the doc generation problem. I'll look into this too (but if you figure it out first, that would be great!).   I'm really sort of hating all of `emcee` at the moment.

File: requirements-dev.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (reneeotten) [reply requirements-dev.txt]: yes, also true.... we should probably add though that we require >3 in ```doc/installation.rst```

File: tests/test_parameter.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reneeotten) [reply tests/test_parameter.py]: Good. We're still adding a warning about the upcoming deprecation now and will remove it in the next+1 release; until it's completely gone we should keep the test.

**Comment by reneeotten** [comment]
> I don't really understand the doc generation problem. I'll look into this too (but if you figure it out first, that would be great!). I'm really sort of hating all of `emcee` at the moment.

Alright... why don't you make the changes discussed above. After that's done I will look at the documentation and get it to pass Travis CI. I will look into ```emcee``` version 3 a bit more myself and do some testing. Let's leave this PR open for a bit while we work out the kinks.

**Comment by reneeotten** [comment]
@newville I am not sure I agree with the more recent changes you made.
- I do think we should deprecate the ```isParameter``` function as discussed in an earlier PR (and here); sure it is not harmful but that's not really an argument for keeping it. I don't think it has a place in ```parameters.py``` and would like to  clean-up the codebase as much as possible now.
- ```Sphinx``` should halt on warnings and stop the build, otherwise SyntaxErrors that cause the documentation not to build correctly can slip through unnoticed as happened with the PR by @Tillsten. Instead, we should fix the build warnings and make sure there are none.

As I said before I will look a bit more into ```emcee``` and can fix the documentation issues if you want.

**Comment by newville** [comment]
@reneeotten  Oh, OK:  just to be clear:

1.  keep `isParameter()` deprecated and explicitly test that it works and is deprecated?

I don't have a strong opinion. Someone might be using this. If so, they'll have to use their own one-liner.   I also don't have a strong opinion on how elaborately we test deprecated features. 

2. use Sphinx's "fail immediately on Warning".  

I'm not so sure about this: warnings aren't failures.  But I am willing to alter the makefile to have that be the default and add a make target (`make debug?`) with more verbosity and to keep going on failures (and maybe add a `make strict` to fail on warnings).  FWIW, I would have never figured out the `emcee` docstring bug without "keep going, render as best you can" and then looking at the results.

**Comment by Tillsten** [comment]
@reneeotten I would have tried to build it, it is has only the sphinx makefile and not a sphinx make.bat.

**Comment by newville** [comment]
@reneeotten @Tillsten the docs now build ok.  I think the codecov reports from should  really not be viewed as failures or success.

File: lmfit/minimizer.py [hunk 6]
```diff
@@ -950,19 +956,109 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
...
+                          probability
+            'chi2' - objective function returns a chi2 value
+        is_weighted : bool
+            If `userfcn` returns a vector of residuals then `is_weighted`
+            specifies if the residuals have been weighted by data uncertainties.
+        nan_policy : str, optional
+            Specifies action if `userfcn` returns NaN values. One of:
+                 'raise' - a `ValueError` is raised
+                'propagate' - the values returned from `userfcn` are un-altered
+                'omit' - the non-finite values are filtered
```
Comment (reneeotten) [inline lmfit/minimizer.py]: I know we currently don't show this function in the documentation so it probably doesn't really matter, but for a list to show up correctly, one would need a newline before/after the elements. Also "raise" is currently not aligned with the other elements

File: lmfit/minimizer.py [hunk 7]
```diff
@@ -950,19 +956,109 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
...
+
+        Returns
+        -------
+        lnprob : float
+            Log posterior probability.
+             """
```
Comment (reneeotten) [inline lmfit/minimizer.py]: at the end of the docstring there should be a newline and the `"""` should be left aligned, so:
```
       lnprob : float
            Log posterior probability.

       """
```

File: lmfit/minimizer.py [hunk 8]
```diff
@@ -1140,17 +1208,15 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
             raise NotImplementedError('You must have emcee to use'
```
Comment (reneeotten) [inline lmfit/minimizer.py]: perhaps add here: ".... emcee **version 3*** to use...."

**Comment by reneeotten** [comment]
> @reneeotten Oh, OK: just to be clear:
>
>     1. keep `isParameter()` deprecated and explicitly test that it works and is deprecated?
> 
> 
> I don't have a strong opinion. Someone might be using this. If so, they'll have to use their own one-liner. I also don't have a strong opinion on how elaborately we test deprecated features.

I am sorry @newville if I caused additional confusion, yes that's indeed what I intended. I made this change in the previous PR were we discussed it and from merging that I assumed that you agreed with that. Also, it isn't really related to this ```emcee``` PR anyway, so let's keep it out of here. The whole idea of having a FutureWarning is that if a lot of people tell us that they actually use this function, we can still re-evaluate if for the next release. 

>     1. use Sphinx's "fail immediately on Warning".
> 
> I'm not so sure about this: warnings aren't failures. But I am willing to alter the makefile to have that be the default and add a make target (`make debug?`) with more verbosity and to keep going on failures (and maybe add a `make strict` to fail on warnings). FWIW, I would have never figured out the `emcee` docstring bug without "keep going, render as best you can" and then looking at the results.

I see your point here but I've also seen a variety of Sphinx syntax warnings that actually resulted in "wrong" and/or incomplete documentation, but did not stop the build. Thank you for making the changes you did and adding a debug option to, with that I think we should be able to resolve such issues locally and have the "final" version build without warnings.

File: lmfit/minimizer.py [hunk 9]
```diff
@@ -950,19 +956,109 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
...
-        this method.
+              float_behavior='posterior', is_weighted=True, seed=None, progress=False):
+        """Bayesian sampling of the posterior distribution using the `emcee` Markov
+        Chain Monte Carlo package. The method assumes that the prior is Uniform.
+        You need to have `emcee` installed to use this method.
 
         Parameters
         ----------
-        params : :class:`~lmfit.parameter.Parameters`, optional
+        params : `lmfit.Parameters`, optional
```
[the remark this answers is not in the document]
Comment (newville) [reply lmfit/minimizer.py]: ok

File: lmfit/minimizer.py [hunk 10]
```diff
@@ -950,19 +956,109 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
...
-        r"""Bayesian sampling of the posterior distribution using `emcee`.
-
-        Bayesian sampling of the posterior distribution for the parameters
-        using the `emcee` Markov Chain Monte Carlo package. The method assumes
-        that the prior is Uniform. You need to have `emcee` installed to use
-        this method.
+              float_behavior='posterior', is_weighted=True, seed=None, progress=False):
+        """Bayesian sampling of the posterior distribution using the `emcee` Markov
+        Chain Monte Carlo package. The method assumes that the prior is Uniform.
+        You need to have `emcee` installed to use this method.
```
[the remark this answers is not in the document]
Comment (newville) [reply lmfit/minimizer.py]: ok

**Comment by newville** [comment]
@reneeotten I think I fixed most of these doc issues.

**Comment by reneeotten** [comment]
thanks @newville ! Are you okay with me doing some git-vodoo to reduce this to a smaller number of commits so that the actual changes are clearer? I have a few more small suggestions which I can commit to this PR as well and then we should be able to wrap this up for now and have the support (minus the PTSampler) as before. 

I'll need some more time to look into ```emcee``` and see how and which parts of version 3 we could/should add but that can happen at a later stage.

**Comment by newville** [comment]
@reneeotten Yes, I'm definitely OK with that.  This got sort of messy, but it's fine to think of it as a single unit.  I would be fine with using Github's "squash and merge" or any variation of rebasing.

For adding more `emcee` code here, as with #601, I'm mostly suspicious that `emcee` (or at least "fitting with `emcee`") isn't working very well.   I'm not convinced it ever really did work well.  So I sort of wonder if there is some problem in the conversion of residual/penalty to log-posterior or something.  I don't see it. I also think there is some degree of "snake-oil" in the one and only `emcee` example.  I definitely agree that MCMC has merit and value, but it seems like something (either our code or `emcee` itself) might not be working correctly.

**Comment by newville** [comment]
@reneeotten Is it OK to (squash and) merge this in a couple of days?  I think we need to stay on track for getting to the next release by the end of the year.

**Comment by reneeotten** [comment]
> @reneeotten Is it OK to (squash and) merge this in a couple of days? I think we need to stay on track for getting to the next release by the end of the year.

I have locally already reorganized your commits and added some of my additional changes. One thing I am looking into still is Issue #590, it looks like you tried to address that in this PR as well. But it doesn't work as intended yet (in my opinion one should be able to add the ```emcee``` method in ```test_itercb.py``` and that should work, but it throws an error right now); I'll try to figure that out tonight or tomorrow and then push to this PR. You can then check my additional commits and if you agree we can squash some of them and merge.

**Comment by newville** [comment]
@reneeotten I believe this PR does address #590.  Specifically, adding an `iter_cb` to `examples/doc_fitting_emcee.py` works for me.

**Comment by reneeotten** [comment]
> @reneeotten I believe this PR does address #590. Specifically, adding an `iter_cb` to `examples/doc_fitting_emcee.py` works for me.

Ehm, perhaps that does depend on the definition of "works for me"; if I add / change the following in ```examples/doc_fitting_emcee.py```:
```
def per_iteration(pars, iter, resid, *args, **kws):
    """Iteration callback, will abort at iteration 23."""
    return iter == 23

<snip>

res = lmfit.minimize(residual, method='emcee', nan_policy='omit', burn=300,
                     steps=1000, thin=20, params=mi.params, is
```

I indeed get a ```lmfit.minimizer.AbortFitException: fit aborted by user.``` but the associated error/Traceback (see below) shouldn't happen. At least for all the other methods we catch the error, set some attributes but do finish normally; or do you not see this?

```emcee: Exception while calling your likelihood function:
  params: [ 2.98644546 -4.3352576   1.31005392 11.81031403 -2.30258148]
  args: (<function residual at 0x10d20ea70>, Parameters([('a1', <Parameter 'a1', value=2.9864454619609777, bounds=[-inf:inf]>), ('a2', <Parameter 'a2', value=-4.33525760330032, bounds=[-inf:inf]>), ('t1', <Parameter 't1', value=1.3100539214083753, bounds=[-inf:inf]>), ('t2', <Parameter 't2', value=11.810314033213828, bounds=[-inf:inf]>), ('__lnsigma', <Parameter '__lnsigma', value=-2.3025814820544053, bounds=[-6.907755278982137:0.6931471805599453]>)]), ['a1', 'a2', 't1', 't2', '__lnsigma'], array([[       -inf,         inf],
       [       -inf,         inf],
       [       -inf,         inf],
       [       -inf,         inf],
       [-6.90775528,  0.69314718]]))
  kwargs: {'is_weighted': False, 'float_behavior': 'posterior', 'userargs': [], 'userkws': {}, 'nan_policy': 'omit'}
  exception:
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/emcee/ensemble.py", line 542, in __call__
    return self.f(x, *self.args, **self.kwargs)
  File "/Users/renee/Library/Python/3.7/lib/python/site-packages/lmfit-0.9.14+46.ge6c0800-py3.7.egg/lmfit/minimizer.py", line 1025, in _lnprob
    raise AbortFitException("fit aborted by user.")
lmfit.minimizer.AbortFitException: fit aborted by user.
Traceback (most recent call last):
  File "doc_fitting_emcee.py", line 55, in <module>
    progress=False, iter_cb=per_iteration)
  File "/Users/renee/Library/Python/3.7/lib/python/site-packages/lmfit-0.9.14+46.ge6c0800-py3.7.egg/lmfit/minimizer.py", line 2388, in minimize
  File "/Users/renee/Library/Python/3.7/lib/python/site-packages/lmfit-0.9.14+46.ge6c0800-py3.7.egg/lmfit/minimizer.py", line 2171, in minimize
  File "/Users/renee/Library/Python/3.7/lib/python/site-packages/lmfit-0.9.14+46.ge6c0800-py3.7.egg/lmfit/minimizer.py", line 1339, in emcee
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/emcee/ensemble.py", line 381, in run_mcmc
    for results in self.sample(initial_state, iterations=nsteps, **kwargs):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/emcee/ensemble.py", line 282, in sample
    state.log_prob, state.blobs = self.compute_log_prob(state.coords)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/emcee/ensemble.py", line 424, in compute_log_prob
    map_func(self.log_prob_fn, (p[i] for i in range(len(p))))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/emcee/ensemble.py", line 542, in __call__
    return self.f(x, *self.args, **self.kwargs)
  File "/Users/renee/Library/Python/3.7/lib/python/site-packages/lmfit-0.9.14+46.ge6c0800-py3.7.egg/lmfit/minimizer.py", line 1025, in _lnprob
lmfit.minimizer.AbortFitException: fit aborted by user.```

**Comment by newville** [comment]
@reneeotten Well, it is progress in that the iteration callback did run.  Maybe we should be handling that abort request better ;).

**Comment by reneeotten** [comment]
@newville I don't think this ```AbortFitException``` is going to work... even with your latest changes it still throws the same error. It's caused in our ```_lnprob``` function and somewhere within the ```emcee``` code ("emcee: Exception while calling your likelihood function"). I other words, we can catch the exception in our code, but already earlier on it throws an exception in the emcee code. 

To be honest I don't think it's worthwhile -at least for now- to try and solve this right now (I am not sure whether it's all that useful in the context of ```emcee``` anyway). What about we drop the ```iter_cb``` stuff for now and get it working without first?

**Comment by newville** [comment]
@reneeotten  Well, `iter_cb` does work with `emcee` to monitor and/or report progress.
I also think the AbortFitException is working.  That is,  with `Minimimizer.emcee()` doing

    try:
        output = self.sampler.run_mcmc(p0, steps, progress=progress)
        self._lastpos = output.coords
    except AbortFitException:
        ....

returning `True` from the iteration callback does cause the "fit" (if that's what we're calling the task that `run_mcmc()` does) to stop.   It appears to me that `emcee.run_mcmc()` *prints* part of a traceback before passing the exception from `Minimizer._lnprob` back to `Minimizer.emcee()`, but the exception is handled in `Minimizer.emcee` instead of propagating to the user.  The user's code calling `Minimizer.emcee()` does not get an exception but a result with `success=False` instead.  The internal state of the "fit" (sampler, chains, etc) are left in a poor and probably unusable state.  Then again, we don't promise that aborting a fit with `iter_cb` will ever give useful results.

This is seeming like an awful lot of effort to put into a fitting method that basically does not work.  If we cannot merge this soon and get to 0.9.15 and 1.0 soon, I believe we will be better off dropping all support for `emcee`.

**Comment by reneeotten** [comment]
@newville okay, I added a few commits. Please let me know what you think about those and if you agree, I will reorganize the commits a bit and we should be good to merge it for now. We can deal with possible remaining things later.

**Comment by newville** [comment]
@reneeotten Yes, and Thanks -- those all look good to me.  FWIW I have no strong opinion on the default value for progress-with-tqdm.

**Comment by reneeotten** [comment]
@newville I've added my commits and squashed the commits a few days ago. As I said I am fine with merging this and we can figure out issues if/when they come up.

**Comment by newville** [comment]
@reneeotten  OK, thanks!  I'll merge this now.

**Pull request merged**